worker/changestream/worker.go

```diff
@@ -0,0 +1,121 @@
+	"github.com/juju/worker/v3"
+	"github.com/juju/worker/v3/catacomb"
+
+	"github.com/juju/juju/core/changestream"
+	"github.com/juju/juju/worker/dbaccessor"
+)
+
+// DBGetter describes the ability to supply a sql.DB
+// reference for a particular database.
+type DBGetter = dbaccessor.DBGetter
+
+// ChangeStream represents a stream of changes that flows from the underlying
+// change log table in the database.
+type ChangeStream interface {
+	// Changes returns a channel for a given namespace. The channel will return
+	// change events that represent change log rows from the database. The
+	// change events will be monotically increasing events (monotonic in that
+	// they're always increasing, but not in that they're always sequential).
+	// The change events will be ordered by the row id and will be coalesced
+	// into a single change event if they are for the same entity and for the
+	// change type.
+	Changes(namespace string) (<-chan changestream.ChangeEvent, error)
+}
+
+// DBStream is the interface that the worker uses to interact with the raw
+// database stream. This is not namespaced and works exactly on the raw
+// database.
+type DBStream interface {
+	worker.Worker
+	Changes() <-chan changestream.ChangeEvent
+}
+
+// WorkerConfig encapsulates the configuration options for the
+// changestream worker.
+type WorkerConfig struct {
+	DBGetter  DBGetter
+	Clock     clock.Clock
+	Logger    Logger
+	NewStream StreamFn
+}
+
+// Validate ensures that the config values are valid.
+func (c *WorkerConfig) Validate() error {
+	if c.Clock == nil {
+		return errors.NotValidf("missing clock")
+	}
+	if c.Logger == nil {
+		return errors.NotValidf("missing logger")
+	}
+
+	return nil
+}
+
+type changeStreamWorker struct {
+	cfg      WorkerConfig
+	catacomb catacomb.Catacomb
+}
+
+func newWorker(cfg WorkerConfig) (*changeStreamWorker, error) {
+	var err error
+	if err = cfg.Validate(); err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	w := &changeStreamWorker{
+		cfg: cfg,
+	}
+
+	if err = catacomb.Invoke(catacomb.Plan{
+		Site: &w.catacomb,
+		Work: w.loop,
+	}); err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	return w, nil
+}
+
+func (w *changeStreamWorker) loop() (err error) {
+	for {
+		select {
+		case <-w.catacomb.Dying():
+			return w.catacomb.ErrDying()
+		}
+	}
+}
+
+// Kill is part of the worker.Worker interface.
+func (w *changeStreamWorker) Kill() {
+	w.catacomb.Kill(nil)
+}
+
+// Wait is part of the worker.Worker interface.
+func (w *changeStreamWorker) Wait() error {
+	return w.catacomb.Wait()
+}
+
+// Changes returns a channel containing all the change events for the given
+// namespace.
+func (w *changeStreamWorker) Changes(namespace string) (<-chan changestream.ChangeEvent, error) {
+	db, err := w.cfg.DBGetter.GetDB(namespace)
+	if err != nil {
+		return nil, errors.Annotatef(err, "getting db for namespace %q", namespace)
+	}
+
+	// TODO (stickupkid): We could potentially cache the streams here and hand
```

Comment:
We will have to do this anyway, because we must monitor the namespaces we have. If a model dies, we need to remove its stream.

Reply:
So this will require some sort of life watcher for models, similar to the state worker. I wonder if there is a nice way to do this, so that we're not tying a dqlite db to mongo db.